builder/llpcBuilderImplSubgroup.cpp

```diff
@@ -0,0 +1,1185 @@
+    }
+    else if (pType->getPrimitiveSizeInBits() == 64)
+    {
+        SmallVector<Value*, 4> bitCasts;
+
+        for (Value* const pArgument : mappedArgs)
+        {
+            bitCasts.push_back(CreateBitCast(pArgument, VectorType::get(getInt32Ty(), 2)));
+        }
+
+        Value* pResult = UndefValue::get(bitCasts[0]->getType());
+
+        for (unsigned i = 0; i < 2; i++)
+        {
+            SmallVector<Value*, 4> elements;
+
+            for (Value* const pBitCast : bitCasts)
+            {
+                elements.push_back(CreateExtractElement(pBitCast, i));
+            }
+
+            Value* const pMap = CreateMapToI32(function, elements, passthroughArgs);
+
+            pResult = CreateInsertElement(pResult, pMap, i);
+        }
+
+        return CreateBitCast(pResult, pType);
+    }
+    else if (pType->isFloatingPointTy())
+    {
+        SmallVector<Value*, 4> elements;
+
+        for (Value* const pArgument : mappedArgs)
+        {
+            elements.push_back(CreateBitCast(pArgument, getIntNTy(pArgument->getType()->getPrimitiveSizeInBits())));
+        }
+
+        Value* const pMap = CreateMapToI32(function, elements, passthroughArgs);
+        return CreateBitCast(pMap, pType);
+    }
+    else if (pType->isIntegerTy(32))
+    {
+        return function(*this, mappedArgs, passthroughArgs);
+    }
+    else
+    {
+        LLPC_NOT_IMPLEMENTED();
+        return nullptr;
+    }
+}
+
+// =====================================================================================================================
+// Create a thread mask for the current thread, a 64-bit integer with a single bit representing the ID of the thread
+// set to 1.
+Value* BuilderImplSubgroup::CreateThreadMask()
+{
+    Value* const pMbcnt = CreateMbcnt(getInt64(UINT64_MAX));
+    Value* const pZExt = CreateZExtOrTrunc(pMbcnt, getInt64Ty());
+    return CreateShl(getInt64(1), pZExt);
+}
+
+// =====================================================================================================================
+// Create a masked operation - taking a thread mask and a mask to and it with, select between x and y only if the
+// current thread is active.
+Value* BuilderImplSubgroup::CreateMaskedOperation(
+    Value* const pThreadMask, // [in] The thread mask, must come from a call to CreateThreadMask. Must be an i64 type.
+    uint64_t     mask,        // The mask to and with the thread mask, must be an i64 type.
+    Value* const pX,          // [in] The x value.
+    Value* const pY)          // [in] The y value.
+{
+    return CreateSelect(CreateICmpNE(CreateAnd(pThreadMask, getInt64(mask)), getInt64(0)), pX, pY);
+}
+
+// =====================================================================================================================
+// Turn a per-lane value (in a VGPR) into a subgroup-wide shared SGPR. A non-zero value contributed by an lane is
+// interpreted as true, otherwise false.
+Value* BuilderImplSubgroup::CreateVGPRToSGPR(
```

Comment:
How about change this function name to CreateBooleanToSgpr or CreateBooleanToBitMask? @amdrexu 
CreateVGPRToSGPR is easy to confuse  with readfirstlane

Reply:
Yes, I will think about this when reviewing the commit.